admin/cleanup.py

```diff
@@ -0,0 +1,379 @@
+            volumes.extend(driver.list_volumes())
+        return volumes
+
+    def _get_cluster_id(self, volume):
+        """
+        Extract the Flocker-specific cluster identifier from the given volume.
+
+        :raise: ``KeyError`` if the given volume is not tagged with a Flocker
+            cluster id.
+        """
+        return _get_tag(volume, 'flocker-cluster-id')
+
+    def _is_test_cluster(self, cluster_id):
+        """
+        Determine whether or not the given Flocker cluster identifier belongs
+        to a cluster created by a test suite run.
+
+        :return: ``True`` if it does, ``False`` if it does not.
+        """
+        try:
+            return UUID(cluster_id).node == self.marker
+        except:
+            err(None, "Could not parse cluster_id {!r}".format(cluster_id))
+            return False
+
+    def _is_test_volume(self, volume):
+        """
+        Determine whether or not the given volume belongs to a test-created
+        Flocker cluster (and is therefore subject to automatic destruction).
+
+        :return: ``True`` if it does, ``False`` if it does not.
+        """
+        try:
+            cluster_id = self._get_cluster_id(volume)
+        except KeyError:
+            return False
+        return self._is_test_cluster(cluster_id)
+
+    def _filter_test_volumes(self, maximum_age, volumes):
+        """
+        From the given list of volumes, find volumes created by tests which are
+        older than the maximum age.
+
+        :param timedelta maximum_age: The oldest a volume may be without being
+            considered old enough to include in the result.
+        :param list all_volumes: The libcloud volumes representing all the
+            volumes we can see on a particular cloud service.
+
+        :rtype: ``VolumeActions``
+        """
+        now = datetime.now(tz=tzutc())
+        destroy = []
+        keep = []
+        for volume in volumes:
+            created = _get_volume_creation_time(volume)
+            if self._is_test_volume(volume) and now - created > maximum_age:
+                destroy.append(volume)
+            else:
+                keep.append(volume)
+        return VolumeActions(destroy=destroy, keep=keep)
+
+    def _destroy_cloud_volumes(self, volumes):
+        """
+        Unconditionally and irrevocably destroy all of the given cloud volumes.
+        """
+        for volume in volumes:
+            try:
+                volume.destroy()
+            except:
+                err(None, "Destroying volume.")
+
+
+def _dumps(obj):
+    """
+    JSON encode an object using some visually pleasing formatting.
+    """
+    return json.dumps(obj, sort_keys=True, indent=4, separators=(',', ': '))
+
+
+def _format_time(when):
+    """
+    Format a time to ISO8601 format.  Or if there is no time, just return
+    ``None``.
+    """
+    if when is None:
+        return None
+    return datetime.isoformat(when)
+
+
+def _get_tag(volume, tag_name):
```

Comment:
I think you actually just want `_get_flocker_cluster_id(volume)`.

I don't think we want to commit to every cloud having a concept of "tags", that tags are key:value pairs, or that flocker-cluster-id will be the key we will use for this tag.

For instance, on GCE tags are more like flags (just a set of strings that match `/[a-z]([a-z0-9-]{0,62}[a-z0-9])?/` you can flag an instance with), and thus are not used for storing cluster id in the current implementation. So instead we set the free-from description field to a JSON blob that we parse to get the cluster id.